File: euler_054.py
from time import time

def create_hands():
    player1 = []
    player2 = []
    with open("p054_poker.txt", "r") as f:
        for line in f.readlines():
            line = line.replace("\n", "")
            line = line.split(" ")
            player1.append(line[:5])
            player2.append(line[5:])
    return (player1, player2)

def value(c):
    if c[0] in ["T", "J", "Q", "K", "A"]:
        return ["T", "J", "Q", "K", "A"].index(c[0]) + 10
    return int(c[0])

def values(h):
    return [value(c) for c in h]

def suits(h):
    return [c[1] for c in h]

def evaluate_hand(h):
    v = values(h)
    cts = [v.count(val) for val in v]
    s = suits(h)
    scores = [0, 0, 0, 0, 0, 0, 0, 0, 0]
    highest = sorted(v, reverse=True)

    #One pair.
    if sorted(cts) == [1, 1, 1, 2, 2]:
        for val in v:
            if v.count(val) == 2:
                scores[8] = val
    else:
        scores[8] = 0

    #Two pairs.
    if sorted(cts) == [1, 2, 2, 2, 2]:
        m = 0
        for val in v:
            if v.count(val) == 2 and val > m:
                m = val
        scores[7] = m
    else:
        scores[7] = 0

    #Three of a kind.
    if sorted(cts) == [1, 1, 3, 3, 3]:
        for val in v:
            if v.count(val) == 3:
                scores[6] = val
    else:
        scores[6] = 0

    #Straight.
    if sorted(cts) == [1, 1, 1, 1, 1] and max(v)%14 - min(v) == 4:
        scores[5] = max(v)
    else:
        scores[5] = 0

    #Flush. Also includes Straight Flush and Royal Flush.
    if len(s) == 1:
        scores[4] = 1
    else:
        scores[4] = 0

    #Full House.
    if sorted(cts) == [2, 2, 3, 3, 3]:
        m = 0
        for val in v:
            if v.count(val) == 3:
                m = val
        scores[3] = m
    else:
        scores[3] = 0

    #Four of a kind.
    if sorted(cts) == [1, 4, 4, 4, 4]:
        m = 0
        for val in v:
            if v.count(val) == 4:
                m = val
        scores[2] = m
    else:
        scores[2] = 0

    #Straight Flush.
    if sorted(cts) == [1, 1, 1, 1, 1] and max(v)%14 - min(v) == 4 and len(s) == 1:
        scores[1] = max(v)
    else:
        scores[1] = 0

    #Royal Flush.
    if sorted(cts) == [1, 1, 1, 1, 1] and max(v) == 14 and len(s) == 1:
        scores[0] = 1
    else:
        scores[0] = 0

    return (scores, highest)

def compare_hands(h1, h2):
    e1 = evaluate_hand(h1)
    e2 = evaluate_hand(h2)
    winner = 0
    for i in range(9):
        if e1[0][i] == e2[0][i]:
            continue
        else:
            if e1[0][i] > e2[0][i]:
                winner = 1
            else:
                winner = 2
            break

    if winner != 0:
        return winner
    else:
        for i in range(5):
            if e1[1][i] == e2[1][i]:
                continue
            else:
                if e1[1][i] > e2[1][i]:
                    winner = 1
                else:
                    winner = 2
                break

    return winner


start = time()

p = create_hands()
p1, p2 = p[0], p[1]

p1_wins = 0
for i in range(len(p1)):
    if compare_hands(p1[i], p2[i]) == 1:
        p1_wins += 1

print("Player 1 wins " + str(p1_wins) + " hands")

stop = time()
print("Time: " + str(stop-start))

File: euler_054.js
const start = Date.now();

const { readFileSync, promises: fsPromises } = require('fs');

function syncReadFile(filename) {
    const contents = readFileSync(filename, 'utf-8');

    const arr = contents.split("\n");

    return arr;
}

// An array of ten cards
const hands = syncReadFile('./euler_054.txt');

// Arrays of 5 cards
let playerOneHands = [];
let playerTwoHands = [];

for (const twoHands of hands) {
    const cards = twoHands.split(" ");
    playerOneHands.push(cards.slice(0,5));
    playerTwoHands.push(cards.slice(5));
}


// Conversion of faces and suits to values
const suits = new Map();
suits.set('S', 1);
suits.set('H', 2);
suits.set('C', 3);
suits.set('D', 4);

const faces = new Map();
for (let v = 2; v < 10; v++) {
    faces.set(v.toString(), v);
}
faces.set("T", 10);
faces.set("J", 11);
faces.set("Q", 12);
faces.set("K", 13);
faces.set("A", 14);

// The (value) signature of a hand
const signature = (hand) => {
    let sig = Array(15);
    sig.fill(0);

    for (const card of hand) {
        sig[faces.get(card[0])]++;
    }
    sig[1] = sig[14]; // Aces can be smallest and largest
    return sig;
}
// End signature

// Test for flush
const flush = (hand) => {
    let s = new Set();
    for (const card of hand) {
        s.add(suits.get(card[1]));
    }
    return s.size === 1;
}
// End test flush

// Test for straight
const straight = (hand) => {
    const sig = signature(hand);
    for (let i = 1; i < 11; i++) {
        if (sig.slice(i, i + 5).map(x => x.toString()).join("") === "11111") {
            return true;
        }
    }
    return false;
}
// End test straight

// Get high card
const highCard = (hand) => {
    let hc = 0;
    for (const card of hand) {
        hc = Math.max(faces.get(card[0]), hc);
    }
    return hc;
}
// End get high card

// Ranking signature of a hand
const ranking = (hand) => {
    let s = signature(hand);
    let r = new Map();
    for (rank of ['RF', 'SF', '4K', 'FH', 'F', 'S', '3K', '2P', '1P']) {
        r.set(rank, 0);
    }
    r.set('HC', highCard(hand));

    // Check for Royal Flush and Straight Flush
    if (flush(hand) && straight(hand)) {
        if (r.get('HC') === 14) {
            r.set('RF', 14);
        } else {
            r.set('SF', r.get('HC'));
        }
    }

    // Check for 4 of a kind
    if (s.some(x => x === 4)) {
        r.set('4K', s.indexOf(4));
    }

    // Check for Full House and three of a kind
    if (s.some(x => x === 3)) {
        if (s.some(x => x === 2)) {
            r.set('FH', [s.indexOf(3), s.indexOf(2)]);
        } else {
            r.set('3K', s.indexOf(3));
        }
    }

    // Check for Flush
    if (flush(hand)) {
        r.set('F', 1);
    }

    // Check for Straight
    if (straight(hand)) {
        r.set('S', r.get('HC'));
    }

    // Check for Two Pairs and One Pair
    let sa = [...s];
    sa[1] = 0;
    if (sa.some(x => x === 2)) {
        let firstPair = sa.indexOf(2);
        if (sa.slice(firstPair + 1).some(x => x === 2)) {
            let lastPair = sa.lastIndexOf(2);
            r.set('2P', [lastPair, firstPair]);
        } else {
            r.set('1P', firstPair);
        }
    }

    return r;
}
// End of ranking

// Compare two hands
const compare = (hand1, hand2) => {
    const r1 = ranking(hand1);
    const r2 = ranking(hand2);

    if (r1.get('RF') !== r2.get('RF')) {
        if (r1.get('RF') > r2.get('RF')) {
            //console.log("RF 1: ", hand1, hand2);
            return 1;
        } else {
            //console.log("RF 2: ", hand1, hand2);
            return 2;
        }
        
    }

    if (r1.get('SF') !== r2.get('SF')) {
        if (r1.get('SF') > r2.get('SF')) {
            //console.log("SF 1: ", hand1, hand2);
            return 1;
        } else {
            //console.log("SF 2: ", hand1, hand2);
            return 2;
        }
    }

    if (r1.get('4K') !== r2.get('4K')) {
        if (r1.get('4K') > r2.get('4K')) {
            //console.log("4K 1: ", hand1, hand2);
            return 1;
        } else {
            //console.log("4K 2: ", hand1, hand2);
            return 2;
        }
    }

    if (r1.get('FH') !== 0 || r2.get('FH') !== 0) {
        if (r1.get('FH')[0] > r2.get('FH')[0]) {
            //console.log("FH 1: ", hand1, hand2);
            return 1;
        }
        if (r1.get('FH')[0] < r2.get('FH')[0]) {
            //console.log("FH 2: ", hand1, hand2);
            return 2;
        }
        if (r1.get('FH')[1] > r2.get('FH')[1]) {
            //console.log("FH 1: ", hand1, hand2);
            return 1;
        } else {
            //console.log("FH 2: ", hand1, hand2);
            return 2;
        }
    }

    if (r1.get('F') !== 0 || r2.get('F') !== 0) {
        if (r1.get('F') > r2.get('F')) {
            //console.log("F 1: ", hand1, hand2);
            return 1;
        }
        if (r1.get('F') < r2.get('F')) {
            return 2;
        }
    }

    if (r1.get('S') !== r2.get('S')) {
        if (r1.get('S') > r2.get('S')) {
            //console.log("S 1: ", hand1, hand2);
            return 1;
        } else {
            //console.log("S 2: ", hand1, hand2);
            return 2;
        }
    }

    if (r1.get('3K') !== r2.get('3K')) {
        if (r1.get('3K') > r2.get('3K')) {
            //console.log("3K 1: ", hand1, hand2);
            return 1;
        } else {
            //console.log("3K 2: ", hand1, hand2);
            return 2;
        }
    }

    if (r1.get('2P') === 0 && r2.get('2P') !== 0) {
        return 2;
    }
    if (r1.get('2P') !== 0 && r2.get('2P') === 0) {
        return 1;
    }
    if (r1.get('2P') !== 0 && r2.get('2P') !== 0) {
        if (r1.get('2P')[0] !== r2.get('2P')[0]) {
            if (r1.get('2P')[0] > r2.get('2P')[0]) {
                return 1;
            } else {
                return 2;
            }
        } else {
            if (r1.get('2P')[1] > r2.get('2P')[1]) {
                return 1;
            } else {
                return 2;
            }
        }
    }

    if (r1.get('1P') !== r2.get('1P')) {
        if (r1.get('1P') > r2.get('1P')) {
            return 1;
        } else {
            return 2;
        }
    }

    if (r1.get('HC') !== r2.get('HC')) {
        if (r1.get('HC') > r2.get('HC')) {
            return 1;
        } else {
            return 2;
        }
    }

    return 0;
}

let playerOneWins = 0

for (let i = 0; i < 1000; i++) {
    if (compare(playerOneHands[i], playerTwoHands[i]) === 1) {
        playerOneWins++;
    }
}

console.log(playerOneHands.length);
console.log(playerTwoHands.length);

const end = Date.now();

console.log("Execute time: ", end - start, " Player 1 wins ", playerOneWins, " hands.");